### PR TITLE
Update template to use ES 5.x mapping

### LINKED ISSUE #386: Mapping changes in Elasticsearch 5.0 for string field type
Elasticsearch is changing `string` field to split them into two variants: `keyword` and `text`. More info about this here: https://github.com/elastic/elasticsearch/issues/12394

The LS template should be changed to reflect this. `string_fields` rule in template should be split to `text_fields` and `keyword_fields` rule. All the `fields.raw` should be updated to `keyword` type and `@message` should be `text`. 

`@version` should be `keyword` too. 

With this change, LS will be locked with ES version, even with HTTP protocol because this change is not backward compatible.

### PR DESCRIPTION
**Pull request by suyograo**
Fixes #386

### CODE REVIEW

**Comment by untergeek** [comment]
LGTM™.  Let's see if the tests pass 😄

**Comment by untergeek** [comment]
Super excited about this, btw.  Definitely going to tweak some upgraders, but +1 for the clean start.

**Comment by suyograo** [comment]
hmm, I have to use some rspec tag trickery to fix this.

**Comment by jordansissel** [comment]
Let's block this until https://github.com/elastic/elasticsearch/issues/19812 is resolved.

**Comment by suyograo** [comment]
Aye. Stalling

**Comment by jordansissel** [comment]
A few of us (@suyograo, @andrewvc, @talevy, etc) had discussions about this offline, and there were two main areas for discussion about the `keyword` name:

1. The `keyword` name in this template was chosen to be consistent with Elasticsearch's use of the word. Further, the name `keyword` is maybe not the best word.
2. The change from `raw` to `keyword` may negatively impact users who upgrade LS from 2.x to 5.x.

FOr the first part, we are discussing in https://github.com/elastic/elasticsearch/issues/19812.

For the second part, we might leave this field named as `raw` to relieve stress on the upgrade experience. We may move to be consistent with Elasticsearch at a later date after we figure out how to minimize pains of upgrades.

**Comment by untergeek** [comment]
In all honesty, I'd multi-field and have both raw and keyword.  This forces users to adjust the template themselves if they're space conscious, and gives us both reverse and forward compatibility, should Kibana standardize on `.keyword` in future releases.

**Comment by suyograo** [comment]
> In all honesty, I'd multi-field and have both raw and keyword. 

I am -1 on this. I wouldn't want users to incur extra disk space because of this. This multi-field is for every string, which is quite a bit.

**Comment by jordansissel** [comment]
Did some historical digging. Logstash v1.3.0 (December 2013) is the first version which shipped a template. Changing `.raw` has the opportunity to break schema with possibly 2.5 years of events stored in Elasticsearch with Logstash.

We discussed this last week somewhere I forget, but we had some thoughts for Logstash 5.0 timeframe:

Premise:

1.By default,  Logstash will only install the template if there is no template already there.
2. We want to ship a template that more closely aligns with Elasticsearch's naming and defaults.
3. Elasticsearch now makes JSON strings dynamically mapped as both `text` and a multifield `.keyword` as keyword mapping type.

Therefore, I summarize a proposal as follows:

1. We change the default template to change `.raw` to `.keyword`
2. Advertise loudly in the changelog and release notes about this change.
3. Document an upgrade path for existing 2.x users to upgrade to the new template, including instructions on how to reindex data to effect a rename of `.raw` to `.keyword`

**Comment by suyograo** [comment]
@jordansissel I've changed this PR to reflect the proposal in https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/462#issuecomment-238376557.  Can you take a look please? /cc @talevy

File: .travis.yml [hunk 1]
```diff
@@ -3,9 +3,9 @@ sudo: false
 jdk: oraclejdk8
 env:
 - INTEGRATION=false
-- INTEGRATION=true ES_VERSION=2.2.0
 - INTEGRATION=true ES_VERSION=1.7.5
-- INTEGRATION=true ES_VERSION=5.0.0-alpha1
+- INTEGRATION=true ES_VERSION=2.3.4
+- INTEGRATION=true ES_VERSION=5.0.0-alpha4
```
Comment (untergeek) [inline .travis.yml]: Shouldn't this be alpha5 now?

Comment (talevy) [inline .travis.yml]: bump version to `alpha5`?

Comment (suyograo) [inline .travis.yml]: Will do

Comment (jsvd) [inline .travis.yml]: This should be increased to alpha5 since it has been released.

**Comment by untergeek** [comment]
LGTM™

File: .travis.yml (hunk 1 above)
Comment (untergeek) [inline .travis.yml]: Hahahaha.  All of us saw the same thing 👍

File: lib/logstash/outputs/elasticsearch/elasticsearch-template-es5x.json [hunk 2]
```diff
@@ -20,24 +20,23 @@
           "match" : "*",
           "match_mapping_type" : "string",
           "mapping" : {
-            "type" : "string", "index" : "analyzed", "omit_norms" : true,
-            "fielddata" : { "format" : "disabled" },
+            "type" : "text", "norms" : false,
             "fields" : {
-              "raw" : {"type": "string", "index" : "not_analyzed", "ignore_above" : 256}
+              "keyword" : { "type": "keyword" }
```
Comment (talevy) [inline lib/logstash/outputs/elasticsearch/elasticsearch-template-es5x.json]: I thought the idea was to let ES run through its defaults in the case that a stringed field is being dynamically mapped

File: spec/integration/outputs/pipeline_spec.rb [hunk 3]
```diff
@@ -1,6 +1,6 @@
 require_relative "../../../spec/es_spec_helper"
 
-describe "Ingest pipeline execution behavior", :integration => true, :version_5x => true do
+describe "Ingest pipeline execution behavior", :integration => true, :version_greater_than_5x => true do
```
Comment (talevy) [inline spec/integration/outputs/pipeline_spec.rb]: `greater_than`? and equal to?

File: lib/logstash/outputs/elasticsearch/elasticsearch-template-es5x.json (hunk 2 above)
Comment (suyograo) [inline lib/logstash/outputs/elasticsearch/elasticsearch-template-es5x.json]: Yes, but we need `"norms" : false,` to be set, which ES does not by default. That's why I had to explicitly add this rule for strings.

Comment (suyograo) [inline lib/logstash/outputs/elasticsearch/elasticsearch-template-es5x.json]: See 2 lines above.

File: spec/integration/outputs/pipeline_spec.rb (hunk 3 above)
Comment (suyograo) [inline spec/integration/outputs/pipeline_spec.rb]: good call

File: lib/logstash/outputs/elasticsearch/elasticsearch-template-es5x.json (hunk 2 above)
Comment (talevy) [inline lib/logstash/outputs/elasticsearch/elasticsearch-template-es5x.json]: ah, I see. sounds good

**Comment by talevy** [comment]
lgtm

**Pull request merged**